{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Waafle Metadata table preparation**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import necessary libraries\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import os\n",
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Counting the number of HGT events for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(555, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>HGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C16-20348-GH</td>\n",
       "      <td>1408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C16-20197-TZ</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C16-20081-TZ</td>\n",
       "      <td>378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C16-20532-TZ</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C16-20516-TZ</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID   HGT\n",
       "0  C16-20348-GH  1408\n",
       "1  C16-20197-TZ   128\n",
       "2  C16-20081-TZ   378\n",
       "3  C16-20532-TZ   148\n",
       "4  C16-20516-TZ    29"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples=[]\n",
    "HGTs=[]\n",
    "\n",
    "#Iterate trough data dir\n",
    "for sample in os.listdir(\"/home/giacomo/Thesis-Internship/waafle/data\"):\n",
    "    #select lgt files\n",
    "    if \".lgt.tsv\" in sample:\n",
    "        #Read file\n",
    "        with open(os.path.join(\"/home/giacomo/Thesis-Internship/waafle/data\",sample), \"r\") as fn:\n",
    "            #Start counter at -1 to exclude column names\n",
    "            c=-1\n",
    "            #count lines\n",
    "            for line in fn:\n",
    "                c+=1  \n",
    "            #store results\n",
    "            HGTs.append(c)\n",
    "        #remove extension\n",
    "        sample=sample.split(\".\")[0].split(\"_\")[0]\n",
    "        #store sample name\n",
    "        samples.append(sample)\n",
    "\n",
    "#Create a dataframe to store the results of the waafle analysis\n",
    "df_waafle=pd.DataFrame(columns=[\"sampleID\",\"HGT\"])\n",
    "#Fill sample column\n",
    "df_waafle[\"sampleID\"]=samples\n",
    "#Fill HGT column\n",
    "df_waafle[\"HGT\"]=HGTs\n",
    "\n",
    "print(df_waafle.shape)\n",
    "df_waafle.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(555, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>HGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C16-20348-GH</td>\n",
       "      <td>1408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C16-20197-TZ</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C16-20081-TZ</td>\n",
       "      <td>378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C16-20532-TZ</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C16-20516-TZ</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID   HGT\n",
       "0  C16-20348-GH  1408\n",
       "1  C16-20197-TZ   128\n",
       "2  C16-20081-TZ   378\n",
       "3  C16-20532-TZ   148\n",
       "4  C16-20516-TZ    29"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples=[]\n",
    "HGTs=[]\n",
    "\n",
    "#Iterate trough data dir\n",
    "for sample in os.listdir(\"/home/giacomo/Thesis-Internship/waafle/data_qc\"):\n",
    "    #select lgt files\n",
    "    if \".lgt.tsv\" in sample:\n",
    "        #Read file\n",
    "        with open(os.path.join(\"/home/giacomo/Thesis-Internship/waafle/data_qc\",sample), \"r\") as fn:\n",
    "            #Start counter at -1 to exclude column names\n",
    "            c=-1\n",
    "            #count lines\n",
    "            for line in fn:\n",
    "                c+=1  \n",
    "            #store results\n",
    "            HGTs.append(c)\n",
    "        #remove extension\n",
    "        sample=sample.split(\".\")[0].split(\"_\")[0]\n",
    "        #store sample name\n",
    "        samples.append(sample)\n",
    "\n",
    "#Create a dataframe to store the results of the waafle analysis\n",
    "df_waafle=pd.DataFrame(columns=[\"sampleID\",\"HGT\"])\n",
    "#Fill sample column\n",
    "df_waafle[\"sampleID\"]=samples\n",
    "#Fill HGT column\n",
    "df_waafle[\"HGT\"]=HGTs\n",
    "\n",
    "print(df_waafle.shape)\n",
    "df_waafle.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Counting the amount of plasmids and viruses for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(555, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>plasmids</th>\n",
       "      <th>viruses</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C16-20409-GH</td>\n",
       "      <td>423</td>\n",
       "      <td>1746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C16-20563-TZ</td>\n",
       "      <td>666</td>\n",
       "      <td>2375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C16-20625-TZ</td>\n",
       "      <td>374</td>\n",
       "      <td>735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C16-20228-GH</td>\n",
       "      <td>523</td>\n",
       "      <td>1059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C16-20039-GH</td>\n",
       "      <td>277</td>\n",
       "      <td>892</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID  plasmids  viruses\n",
       "0  C16-20409-GH       423     1746\n",
       "1  C16-20563-TZ       666     2375\n",
       "2  C16-20625-TZ       374      735\n",
       "3  C16-20228-GH       523     1059\n",
       "4  C16-20039-GH       277      892"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples1=[]\n",
    "samples2=[]\n",
    "\n",
    "plasmids=[]\n",
    "viruses=[]\n",
    "#Iterate trough data dir\n",
    "for sample in os.listdir(\"/home/giacomo/Thesis-Internship/waafle/mge_data/plasmids\"):\n",
    "    with open(os.path.join(\"/home/giacomo/Thesis-Internship/waafle/mge_data/plasmids\",sample), \"r\") as fn:\n",
    "        #Start counter at -1 to exclude column names\n",
    "        c=-1\n",
    "        #count lines\n",
    "        for line in fn:\n",
    "            c+=1  \n",
    "        #store results\n",
    "        plasmids.append(c)\n",
    "    samples1.append(sample.split(\".\")[0].split(\"_\")[0])\n",
    "           \n",
    "for sample in os.listdir(\"/home/giacomo/Thesis-Internship/waafle/mge_data/viruses\"):\n",
    "    with open(os.path.join(\"/home/giacomo/Thesis-Internship/waafle/mge_data/viruses\",sample), \"r\") as fn:\n",
    "        #Start counter at -1 to exclude column names\n",
    "        v=-1\n",
    "        #count lines\n",
    "        for line in fn:\n",
    "            v+=1  \n",
    "        #store results\n",
    "        viruses.append(v)\n",
    "            \n",
    "    samples2.append(sample.split(\".\")[0].split(\"_\")[0])\n",
    "\n",
    "#Create a dataframe to store the results of the waafle analysis\n",
    "df_plasmids=pd.DataFrame(columns=[\"sampleID\",\"plasmids\"])\n",
    "df_viruses=pd.DataFrame(columns=[\"sampleID\",\"viruses\"])\n",
    "\n",
    "#Fill sample column\n",
    "df_plasmids[\"sampleID\"]=samples1\n",
    "#Fill plasmaids column\n",
    "df_plasmids[\"plasmids\"]=plasmids\n",
    "\n",
    "#Fill sample column\n",
    "df_viruses[\"sampleID\"]=samples2\n",
    "#Fill viruses column\n",
    "df_viruses[\"viruses\"]=viruses\n",
    "\n",
    "#Merge dataframes\n",
    "df_mge=pd.merge(df_plasmids,df_viruses, on=\"sampleID\", how=\"outer\")\n",
    "\n",
    "print(df_mge.shape)\n",
    "df_mge.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load the metadata table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(559, 14)\n",
      "Samples C16-20043-GH C16-20087-GH C16-20128-GH C16-20281-GH C16-20313-GH C16-20314-GH C16-20389-GH C16-20547-TZ have very low sequencing depth and were removed from the analysis.\n",
      "Sample C16-20430-TZ has no sequencing depth information and has no metadata associated so it was removed from the analysis.\n",
      "Sample C16-20547-TZ has no assembled contigs therefore was skipped by both waafle and genomad.\n",
      "Moreover, samples C16-20313-GH C16-20043-GH C16-20430-TZ have no reads or contigs, therefore while waafle assigned 0 HGT events, Genomad couldn't analyze them.\n",
      "(550, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>subjectID</th>\n",
       "      <th>householdID</th>\n",
       "      <th>family_role</th>\n",
       "      <th>species</th>\n",
       "      <th>date</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_days</th>\n",
       "      <th>age_months</th>\n",
       "      <th>age_years</th>\n",
       "      <th>diarrhea_last24h</th>\n",
       "      <th>location</th>\n",
       "      <th>country</th>\n",
       "      <th>Dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C16-20009-TZ</td>\n",
       "      <td>C16-20009-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C16-20010-TZ</td>\n",
       "      <td>C16-20010-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C16-20011-TZ</td>\n",
       "      <td>C16-20011-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C16-20012-TZ</td>\n",
       "      <td>C16-20012-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C16-20013-TZ</td>\n",
       "      <td>C16-20013-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID     subjectID   householdID family_role species      date  \\\n",
       "0  C16-20009-TZ  C16-20009-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "1  C16-20010-TZ  C16-20010-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "2  C16-20011-TZ  C16-20011-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "3  C16-20012-TZ  C16-20012-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "4  C16-20013-TZ  C16-20013-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "\n",
       "   sex  age_days  age_months  age_years diarrhea_last24h location   country  \\\n",
       "0  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "1  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "2  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "3  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "4  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "\n",
       "                    Dataset  \n",
       "0  CM_ghanatanzania_animals  \n",
       "1  CM_ghanatanzania_animals  \n",
       "2  CM_ghanatanzania_animals  \n",
       "3  CM_ghanatanzania_animals  \n",
       "4  CM_ghanatanzania_animals  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata=pd.read_csv(\"/home/giacomo/Thesis-Internship/mdata_GhanaTanzania_animals_humans_seq_20231205.txt\", sep=\"\\t\")\n",
    "print(df_metadata.shape)\n",
    "\n",
    "# Remove rows\n",
    "df_metadata = df_metadata[~df_metadata['sampleID'].isin([\"C16-20043-GH\",\"C16-20087-GH\",\"C16-20128-GH\",\"C16-20281-GH\",\"C16-20313-GH\",\"C16-20314-GH\",\"C16-20389-GH\",\"C16-20430-TZ\",\"C16-20547-TZ\"])]\n",
    "print (\"Samples C16-20043-GH C16-20087-GH C16-20128-GH C16-20281-GH C16-20313-GH C16-20314-GH C16-20389-GH C16-20547-TZ have very low sequencing depth and were removed from the analysis.\")\n",
    "print (\"Sample C16-20430-GH has no sequencing depth information and has no metadata associated so it was removed from the analysis.\")\n",
    "print (\"Sample C16-20547-TZ has no assembled contigs therefore was skipped by both waafle and genomad.\")\n",
    "print(\"Moreover, samples C16-20313-GH C16-20043-GH C16-20430-TZ have no reads or contigs, therefore while waafle assigned 0 HGT events, Genomad couldn't analyze them.\")\n",
    "\n",
    "print(df_metadata.shape)\n",
    "df_metadata.head(5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Integrate info about species richness and sequencing depth"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Richness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(558, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>richness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C16-20002-TZ</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C16-20003-TZ</td>\n",
       "      <td>387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C16-20005-TZ</td>\n",
       "      <td>382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C16-20070-TZ</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C16-20071-TZ</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID  richness\n",
       "0  C16-20002-TZ       248\n",
       "1  C16-20003-TZ       387\n",
       "2  C16-20005-TZ       382\n",
       "3  C16-20070-TZ       163\n",
       "4  C16-20071-TZ       324"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_richness=pd.read_csv(\"/home/giacomo/Thesis-Internship/sample_richness.txt\",sep=\"\\t\")\n",
    "\n",
    "print(df_richness.shape)\n",
    "df_richness.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Depth (remove unsequenced samples, NaN values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples that did not have any sequencing information (NaN entries): C16-20430-TZ\n",
      "(558, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>n_of_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C16-20113-GH</td>\n",
       "      <td>54643887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C16-20126-GH</td>\n",
       "      <td>72428756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C16-20292-GH</td>\n",
       "      <td>31579677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C16-20029-GH</td>\n",
       "      <td>71784691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C16-20245-GH</td>\n",
       "      <td>28717616</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID  n_of_reads\n",
       "0  C16-20113-GH    54643887\n",
       "1  C16-20126-GH    72428756\n",
       "2  C16-20292-GH    31579677\n",
       "3  C16-20029-GH    71784691\n",
       "4  C16-20245-GH    28717616"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_depth=pd.read_csv(\"/home/giacomo/Thesis-Internship/ghanatanzania_stats.txt\",sep=\"\\t\")\n",
    "\n",
    "#Select only the columns of interest\n",
    "df_depth=df_depth[['file_path','n_of_reads']]\n",
    "\n",
    "#From file path extract sample name\n",
    "df_depth['file_path'] = df_depth['file_path'].str.split('/').str[9]\n",
    "\n",
    "#Amount of Nan values\n",
    "df_nan_d=df_depth[\"n_of_reads\"].isnull()\n",
    "\n",
    "#Extract row with Nan values\n",
    "df_nan_d=df_depth[df_nan_d]\n",
    "\n",
    "#Extract sample name \n",
    "df_nan_d=df_nan_d['file_path'].to_string()\n",
    "\n",
    "#Drop NaN values\n",
    "df_depth.dropna(inplace=True)\n",
    "print(\"Samples that did not have any sequencing information (NaN entries):\", df_nan_d.split()[1] )\n",
    "\n",
    "#Transform n_of_reads to integer\n",
    "df_depth['n_of_reads']=df_depth['n_of_reads'].astype(int)\n",
    "\n",
    "#Change column name\n",
    "df_depth.rename(columns={\"file_path\":\"sampleID\"}, inplace=True)\n",
    "\n",
    "print(df_depth.shape)\n",
    "df_depth.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge richness and depth dataframes with metatadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(550, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>subjectID</th>\n",
       "      <th>householdID</th>\n",
       "      <th>family_role</th>\n",
       "      <th>species</th>\n",
       "      <th>date</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_days</th>\n",
       "      <th>age_months</th>\n",
       "      <th>age_years</th>\n",
       "      <th>diarrhea_last24h</th>\n",
       "      <th>location</th>\n",
       "      <th>country</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>richness</th>\n",
       "      <th>n_of_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C16-20009-TZ</td>\n",
       "      <td>C16-20009-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "      <td>541</td>\n",
       "      <td>114685552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C16-20010-TZ</td>\n",
       "      <td>C16-20010-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "      <td>841</td>\n",
       "      <td>65818230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C16-20011-TZ</td>\n",
       "      <td>C16-20011-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "      <td>1270</td>\n",
       "      <td>198092810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C16-20012-TZ</td>\n",
       "      <td>C16-20012-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "      <td>1055</td>\n",
       "      <td>94316346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C16-20013-TZ</td>\n",
       "      <td>C16-20013-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "      <td>51</td>\n",
       "      <td>31387064</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID     subjectID   householdID family_role species      date  \\\n",
       "0  C16-20009-TZ  C16-20009-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "1  C16-20010-TZ  C16-20010-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "2  C16-20011-TZ  C16-20011-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "3  C16-20012-TZ  C16-20012-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "4  C16-20013-TZ  C16-20013-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "\n",
       "   sex  age_days  age_months  age_years diarrhea_last24h location   country  \\\n",
       "0  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "1  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "2  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "3  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "4  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "\n",
       "                    Dataset  richness  n_of_reads  \n",
       "0  CM_ghanatanzania_animals       541   114685552  \n",
       "1  CM_ghanatanzania_animals       841    65818230  \n",
       "2  CM_ghanatanzania_animals      1270   198092810  \n",
       "3  CM_ghanatanzania_animals      1055    94316346  \n",
       "4  CM_ghanatanzania_animals        51    31387064  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata=pd.merge(df_metadata,df_richness, on=\"sampleID\", how=\"inner\")\n",
    "df_metadata=pd.merge(df_metadata,df_depth, on=\"sampleID\", how=\"inner\")\n",
    "\n",
    "print(df_metadata.shape)\n",
    "df_metadata.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge the HGT and MGE dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(550, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>subjectID</th>\n",
       "      <th>householdID</th>\n",
       "      <th>family_role</th>\n",
       "      <th>species</th>\n",
       "      <th>date</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_days</th>\n",
       "      <th>age_months</th>\n",
       "      <th>age_years</th>\n",
       "      <th>diarrhea_last24h</th>\n",
       "      <th>location</th>\n",
       "      <th>country</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>richness</th>\n",
       "      <th>n_of_reads</th>\n",
       "      <th>HGT</th>\n",
       "      <th>plasmids</th>\n",
       "      <th>viruses</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C16-20009-TZ</td>\n",
       "      <td>C16-20009-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "      <td>541</td>\n",
       "      <td>114685552</td>\n",
       "      <td>26</td>\n",
       "      <td>1369</td>\n",
       "      <td>3967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C16-20010-TZ</td>\n",
       "      <td>C16-20010-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "      <td>841</td>\n",
       "      <td>65818230</td>\n",
       "      <td>1037</td>\n",
       "      <td>358</td>\n",
       "      <td>1867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C16-20011-TZ</td>\n",
       "      <td>C16-20011-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "      <td>1270</td>\n",
       "      <td>198092810</td>\n",
       "      <td>2302</td>\n",
       "      <td>1980</td>\n",
       "      <td>9620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C16-20012-TZ</td>\n",
       "      <td>C16-20012-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "      <td>1055</td>\n",
       "      <td>94316346</td>\n",
       "      <td>1799</td>\n",
       "      <td>772</td>\n",
       "      <td>4758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C16-20013-TZ</td>\n",
       "      <td>C16-20013-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Korogwe</td>\n",
       "      <td>Tanzania</td>\n",
       "      <td>CM_ghanatanzania_animals</td>\n",
       "      <td>51</td>\n",
       "      <td>31387064</td>\n",
       "      <td>31</td>\n",
       "      <td>342</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID     subjectID   householdID family_role species      date  \\\n",
       "0  C16-20009-TZ  C16-20009-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "1  C16-20010-TZ  C16-20010-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "2  C16-20011-TZ  C16-20011-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "3  C16-20012-TZ  C16-20012-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "4  C16-20013-TZ  C16-20013-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "\n",
       "   sex  age_days  age_months  age_years diarrhea_last24h location   country  \\\n",
       "0  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "1  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "2  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "3  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "4  NaN       NaN         NaN        NaN              NaN  Korogwe  Tanzania   \n",
       "\n",
       "                    Dataset  richness  n_of_reads   HGT  plasmids  viruses  \n",
       "0  CM_ghanatanzania_animals       541   114685552    26      1369     3967  \n",
       "1  CM_ghanatanzania_animals       841    65818230  1037       358     1867  \n",
       "2  CM_ghanatanzania_animals      1270   198092810  2302      1980     9620  \n",
       "3  CM_ghanatanzania_animals      1055    94316346  1799       772     4758  \n",
       "4  CM_ghanatanzania_animals        51    31387064    31       342      296  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata=pd.merge(df_metadata,df_waafle, on=\"sampleID\", how=\"inner\")\n",
    "df_metadata=pd.merge(df_metadata,df_mge, on=\"sampleID\", how=\"inner\")\n",
    "\n",
    "print(df_metadata.shape)\n",
    "df_metadata.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalize HGT events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=df_metadata.copy()\n",
    "\n",
    "# Normalize HGT events by sequence depth and richness\n",
    "df[\"HGT_richness\"]=df[\"HGT\"]/df[\"richness\"]\n",
    "#df[\"HGT_depth\"]=df[\"HGT\"]/df[\"n_of_reads\"]\n",
    "max_depth=max(df[\"n_of_reads\"])\n",
    "df[\"HGT_depth\"]=df[\"HGT\"]*(max_depth/df[\"n_of_reads\"])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalize MGE events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(550, 25)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>subjectID</th>\n",
       "      <th>householdID</th>\n",
       "      <th>family_role</th>\n",
       "      <th>species</th>\n",
       "      <th>date</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_days</th>\n",
       "      <th>age_months</th>\n",
       "      <th>age_years</th>\n",
       "      <th>...</th>\n",
       "      <th>n_of_reads</th>\n",
       "      <th>HGT</th>\n",
       "      <th>plasmids</th>\n",
       "      <th>viruses</th>\n",
       "      <th>HGT_richness</th>\n",
       "      <th>HGT_depth</th>\n",
       "      <th>plasmids_richness</th>\n",
       "      <th>plasmids_depth</th>\n",
       "      <th>viruses_richness</th>\n",
       "      <th>viruses_depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C16-20009-TZ</td>\n",
       "      <td>C16-20009-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>114685552</td>\n",
       "      <td>26</td>\n",
       "      <td>1369</td>\n",
       "      <td>3967</td>\n",
       "      <td>0.048059</td>\n",
       "      <td>61.937528</td>\n",
       "      <td>2.530499</td>\n",
       "      <td>3261.249085</td>\n",
       "      <td>7.332717</td>\n",
       "      <td>9450.237488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C16-20010-TZ</td>\n",
       "      <td>C16-20010-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>65818230</td>\n",
       "      <td>1037</td>\n",
       "      <td>358</td>\n",
       "      <td>1867</td>\n",
       "      <td>1.233056</td>\n",
       "      <td>4304.490848</td>\n",
       "      <td>0.425684</td>\n",
       "      <td>1486.024806</td>\n",
       "      <td>2.219976</td>\n",
       "      <td>7749.743890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C16-20011-TZ</td>\n",
       "      <td>C16-20011-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>198092810</td>\n",
       "      <td>2302</td>\n",
       "      <td>1980</td>\n",
       "      <td>9620</td>\n",
       "      <td>1.812598</td>\n",
       "      <td>3174.869203</td>\n",
       "      <td>1.559055</td>\n",
       "      <td>2730.773684</td>\n",
       "      <td>7.574803</td>\n",
       "      <td>13267.698406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C16-20012-TZ</td>\n",
       "      <td>C16-20012-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>94316346</td>\n",
       "      <td>1799</td>\n",
       "      <td>772</td>\n",
       "      <td>4758</td>\n",
       "      <td>1.705213</td>\n",
       "      <td>5211.148242</td>\n",
       "      <td>0.731754</td>\n",
       "      <td>2236.245938</td>\n",
       "      <td>4.509953</td>\n",
       "      <td>13782.458774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C16-20013-TZ</td>\n",
       "      <td>C16-20013-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>31387064</td>\n",
       "      <td>31</td>\n",
       "      <td>342</td>\n",
       "      <td>296</td>\n",
       "      <td>0.607843</td>\n",
       "      <td>269.836212</td>\n",
       "      <td>6.705882</td>\n",
       "      <td>2976.902731</td>\n",
       "      <td>5.803922</td>\n",
       "      <td>2576.500609</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID     subjectID   householdID family_role species      date  \\\n",
       "0  C16-20009-TZ  C16-20009-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "1  C16-20010-TZ  C16-20010-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "2  C16-20011-TZ  C16-20011-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "3  C16-20012-TZ  C16-20012-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "4  C16-20013-TZ  C16-20013-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "\n",
       "   sex  age_days  age_months  age_years  ... n_of_reads   HGT plasmids  \\\n",
       "0  NaN       NaN         NaN        NaN  ...  114685552    26     1369   \n",
       "1  NaN       NaN         NaN        NaN  ...   65818230  1037      358   \n",
       "2  NaN       NaN         NaN        NaN  ...  198092810  2302     1980   \n",
       "3  NaN       NaN         NaN        NaN  ...   94316346  1799      772   \n",
       "4  NaN       NaN         NaN        NaN  ...   31387064    31      342   \n",
       "\n",
       "  viruses  HGT_richness    HGT_depth  plasmids_richness  plasmids_depth  \\\n",
       "0    3967      0.048059    61.937528           2.530499     3261.249085   \n",
       "1    1867      1.233056  4304.490848           0.425684     1486.024806   \n",
       "2    9620      1.812598  3174.869203           1.559055     2730.773684   \n",
       "3    4758      1.705213  5211.148242           0.731754     2236.245938   \n",
       "4     296      0.607843   269.836212           6.705882     2976.902731   \n",
       "\n",
       "   viruses_richness  viruses_depth  \n",
       "0          7.332717    9450.237488  \n",
       "1          2.219976    7749.743890  \n",
       "2          7.574803   13267.698406  \n",
       "3          4.509953   13782.458774  \n",
       "4          5.803922    2576.500609  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Normalize plasmids by sequence depth and richness\n",
    "df[\"plasmids_richness\"]=df[\"plasmids\"]/df[\"richness\"]\n",
    "df[\"plasmids_depth\"]=df[\"plasmids\"]*(max_depth/df[\"n_of_reads\"])\n",
    "\n",
    "# Normalize viruses by sequence depth and richness\n",
    "df[\"viruses_richness\"]=df[\"viruses\"]/df[\"richness\"]\n",
    "df[\"viruses_depth\"]=df[\"viruses\"]*(max_depth/df[\"n_of_reads\"])\n",
    "\n",
    "print(df.shape)\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create a comprensive MGE column and normalize it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(550, 28)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>subjectID</th>\n",
       "      <th>householdID</th>\n",
       "      <th>family_role</th>\n",
       "      <th>species</th>\n",
       "      <th>date</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_days</th>\n",
       "      <th>age_months</th>\n",
       "      <th>age_years</th>\n",
       "      <th>...</th>\n",
       "      <th>viruses</th>\n",
       "      <th>HGT_richness</th>\n",
       "      <th>HGT_depth</th>\n",
       "      <th>plasmids_richness</th>\n",
       "      <th>plasmids_depth</th>\n",
       "      <th>viruses_richness</th>\n",
       "      <th>viruses_depth</th>\n",
       "      <th>mge_total</th>\n",
       "      <th>mge_total_richness</th>\n",
       "      <th>mge_total_depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C16-20009-TZ</td>\n",
       "      <td>C16-20009-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3967</td>\n",
       "      <td>0.048059</td>\n",
       "      <td>61.937528</td>\n",
       "      <td>2.530499</td>\n",
       "      <td>3261.249085</td>\n",
       "      <td>7.332717</td>\n",
       "      <td>9450.237488</td>\n",
       "      <td>5336</td>\n",
       "      <td>9.863216</td>\n",
       "      <td>12711.486573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C16-20010-TZ</td>\n",
       "      <td>C16-20010-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1867</td>\n",
       "      <td>1.233056</td>\n",
       "      <td>4304.490848</td>\n",
       "      <td>0.425684</td>\n",
       "      <td>1486.024806</td>\n",
       "      <td>2.219976</td>\n",
       "      <td>7749.743890</td>\n",
       "      <td>2225</td>\n",
       "      <td>2.645660</td>\n",
       "      <td>9235.768696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C16-20011-TZ</td>\n",
       "      <td>C16-20011-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>9620</td>\n",
       "      <td>1.812598</td>\n",
       "      <td>3174.869203</td>\n",
       "      <td>1.559055</td>\n",
       "      <td>2730.773684</td>\n",
       "      <td>7.574803</td>\n",
       "      <td>13267.698406</td>\n",
       "      <td>11600</td>\n",
       "      <td>9.133858</td>\n",
       "      <td>15998.472090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C16-20012-TZ</td>\n",
       "      <td>C16-20012-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4758</td>\n",
       "      <td>1.705213</td>\n",
       "      <td>5211.148242</td>\n",
       "      <td>0.731754</td>\n",
       "      <td>2236.245938</td>\n",
       "      <td>4.509953</td>\n",
       "      <td>13782.458774</td>\n",
       "      <td>5530</td>\n",
       "      <td>5.241706</td>\n",
       "      <td>16018.704712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C16-20013-TZ</td>\n",
       "      <td>C16-20013-TZ</td>\n",
       "      <td>C16-10012-TZ</td>\n",
       "      <td>animal</td>\n",
       "      <td>cow</td>\n",
       "      <td>23/11/16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>296</td>\n",
       "      <td>0.607843</td>\n",
       "      <td>269.836212</td>\n",
       "      <td>6.705882</td>\n",
       "      <td>2976.902731</td>\n",
       "      <td>5.803922</td>\n",
       "      <td>2576.500609</td>\n",
       "      <td>638</td>\n",
       "      <td>12.509804</td>\n",
       "      <td>5553.403340</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID     subjectID   householdID family_role species      date  \\\n",
       "0  C16-20009-TZ  C16-20009-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "1  C16-20010-TZ  C16-20010-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "2  C16-20011-TZ  C16-20011-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "3  C16-20012-TZ  C16-20012-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "4  C16-20013-TZ  C16-20013-TZ  C16-10012-TZ      animal     cow  23/11/16   \n",
       "\n",
       "   sex  age_days  age_months  age_years  ... viruses HGT_richness  \\\n",
       "0  NaN       NaN         NaN        NaN  ...    3967     0.048059   \n",
       "1  NaN       NaN         NaN        NaN  ...    1867     1.233056   \n",
       "2  NaN       NaN         NaN        NaN  ...    9620     1.812598   \n",
       "3  NaN       NaN         NaN        NaN  ...    4758     1.705213   \n",
       "4  NaN       NaN         NaN        NaN  ...     296     0.607843   \n",
       "\n",
       "     HGT_depth plasmids_richness  plasmids_depth  viruses_richness  \\\n",
       "0    61.937528          2.530499     3261.249085          7.332717   \n",
       "1  4304.490848          0.425684     1486.024806          2.219976   \n",
       "2  3174.869203          1.559055     2730.773684          7.574803   \n",
       "3  5211.148242          0.731754     2236.245938          4.509953   \n",
       "4   269.836212          6.705882     2976.902731          5.803922   \n",
       "\n",
       "   viruses_depth  mge_total  mge_total_richness  mge_total_depth  \n",
       "0    9450.237488       5336            9.863216     12711.486573  \n",
       "1    7749.743890       2225            2.645660      9235.768696  \n",
       "2   13267.698406      11600            9.133858     15998.472090  \n",
       "3   13782.458774       5530            5.241706     16018.704712  \n",
       "4    2576.500609        638           12.509804      5553.403340  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"mge_total\"]=df[\"plasmids\"]+df[\"viruses\"]\n",
    "df[\"mge_total_richness\"]=df[\"mge_total\"]/df[\"richness\"]\n",
    "df[\"mge_total_depth\"]=df[\"mge_total\"]*(max_depth/df[\"n_of_reads\"])\n",
    "\n",
    "print(df.shape)\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"/home/giacomo/Thesis-Internship/waafle/waafle_metadata.tsv\", sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
